src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java

```diff
@@ -21,520 +21,550 @@
+         }
+
+         //If the farmer has no currentField and there is no field which needs work, check fields.
+         if (building.getCurrentField() == null && building.getFieldToWorkOn() == null)
+         {
+             building.resetFields();
+             return AIState.IDLE;
+         }
+
+         @Nullable Field currentField = building.getCurrentField();
+
+         if (currentField.needsWork())
+         {
+             switch (currentField.getFieldStage())
+             {
+                 case EMPTY:
+                     if (!checkForHoe())
+                     {
+                         return walkToBlock(currentField.getLocation()) ? AIState.PREPARING : AIState.FARMER_HOE;
+                     }
+                     break;
+                 case HOED:
+                     if (canGoPlanting(currentField, building))
+                     {
+                         return walkToBlock(currentField.getLocation()) ? AIState.PREPARING : AIState.FARMER_PLANT;
+                     }
+                     break;
+                 case PLANTED:
+                     return walkToBlock(currentField.getLocation()) ? AIState.PREPARING : AIState.FARMER_HARVEST;
+                 default:
+                     break;
+             }
+         }
+         else
+         {
+             getOwnBuilding().setCurrentField(null);
+         }
+         return AIState.PREPARING;
+     }
+
+     /**
+      * Searches and adds a field that has not been taken yet for the farmer and then adds it to the list.
+      */
+     private void searchAndAddFields()
+     {
+         Colony colony = worker.getColony();
+         if (colony != null)
+         {
+             @Nullable Field newField = colony.getFreeField(worker.getName());
+
+             if (newField != null && getOwnBuilding() != null)
+             {
+                 getOwnBuilding().addFarmerFields(newField);
+             }
+         }
+     }
+
+     /**
+      * Checks if the farmer is ready to plant.
+      *
+      * @param currentField the field to plant.
+      * @return true if he is ready.
+      */
+     private boolean canGoPlanting(@NotNull Field currentField, @NotNull BuildingFarmer buildingFarmer)
+     {
+         if (currentField.getSeed() == null)
+         {
+             chatSpamFilter.talkWithoutSpam("entity.farmer.noSeedSet");
+             buildingFarmer.setCurrentField(null);
+             return false;
+         }
+
+         if (shouldTryToGetSeed)
+         {
+             if (walkToBuilding())
+             {
+                 return false;
+             }
+             isInHut(new ItemStack(currentField.getSeed()));
+             shouldTryToGetSeed = false;
+         }
+
+         return true;
+     }
+
+     /**
+      * Executes the hoeing of the field.
+      *
+      * @return the next state.
+      */
+     @NotNull
+     private AIState hoe()
+     {
+         @Nullable BuildingFarmer buildingFarmer = getOwnBuilding();
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'buildingFarmer' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)